pkg/urbit/tests/epoc_tests.c

```diff
@@ -0,0 +1,148 @@
+    c3_list_pushb(lis_u, &eve_c, sizeof(eve_c));
+    c3_assert(u3_epoc_commit(poc_u, c3_list_peekb(lis_u), 1));
+    c3_free(c3_list_popb(lis_u));
+    u3_epoc_close(poc_u);
+    c3_free(poc_u);
+
+    // Load recently closed epoch.
+    poc_u = u3_epoc_open(pax_u, NULL);
+    c3_assert(poc_u);
+    c3_assert(!u3_epoc_is_empty(poc_u));
+    c3_assert(1 == u3_epoc_len(poc_u));
+    c3_assert(fir_d == u3_epoc_first_commit(poc_u));
+    c3_assert(fir_d == u3_epoc_last_commit(poc_u));
+    c3_assert(c3_path_eq(pax_u, u3_epoc_path(poc_u)));
+
+    c3_path_free(pax_u);
+  }
+
+  {
+    // Attempt to load epoch that does not exist.
+    c3_path* const pax_u = c3_path_fv(4, "/", "tmp", "event_log", "0i138");
+    u3_epoc*       poc_u = u3_epoc_open(pax_u, NULL);
+    c3_assert(!poc_u);
+    c3_path_free(pax_u);
+  }
+
+  {
+    // Load existing epoch.
+    c3_path* const pax_u = c3_path_fv(4, "/", "tmp", "event_log", "0i137");
+    u3_epoc*       poc_u = u3_epoc_open(pax_u, NULL);
+    c3_assert(poc_u);
+    c3_assert(!u3_epoc_is_empty(poc_u));
+    c3_assert(1 == u3_epoc_len(poc_u));
+    c3_assert(fir_d == u3_epoc_first_commit(poc_u));
+    c3_assert(fir_d == u3_epoc_last_commit(poc_u));
+    c3_assert(c3_path_eq(pax_u, u3_epoc_path(poc_u)));
+
+    // Commit more events to the epoch and then close it.
+    size_t cnt_i = 17;
+    for ( size_t idx_i = 0; idx_i < cnt_i; idx_i++ ) {
+      c3_y tmp_c = eve_c + idx_i + 1;
+      c3_list_pushb(lis_u, &tmp_c, sizeof(tmp_c));
+      c3_assert(u3_epoc_commit(poc_u, c3_list_peekb(lis_u), 1));
+      c3_free(c3_list_popb(lis_u));
+    }
+    u3_epoc_close(poc_u);
+    c3_free(poc_u);
+
+    // Load recently closed epoch.
+    poc_u = u3_epoc_open(pax_u, NULL);
+    c3_assert(poc_u);
+    c3_assert(cnt_i + 1 == u3_epoc_len(poc_u));
+    c3_assert(fir_d == u3_epoc_first_commit(poc_u));
+    c3_assert(fir_d + cnt_i == u3_epoc_last_commit(poc_u));
+    c3_assert(c3_path_eq(pax_u, u3_epoc_path(poc_u)));
+
+    // Iterate over the epoch's events.
+    c3_assert(u3_epoc_iter_open(poc_u, fir_d));
+    for ( size_t idx_i = 0; idx_i < u3_epoc_len(poc_u); idx_i++ ) {
+      c3_y*  tmp_c;
+      size_t len_i;
+      c3_assert(u3_epoc_iter_step(poc_u, &tmp_c, &len_i));
+      c3_assert('a' + idx_i == *tmp_c);
+      c3_assert(sizeof(c3_y) == len_i);
+    }
+    u3_epoc_iter_close(poc_u);
+    u3_epoc_close(poc_u);
+    c3_free(poc_u);
+
+    unlink(c3_path_str(pax_u));
+    c3_path_free(pax_u);
+  }
+  unlink(c3_path_str(par_u));
+  c3_path_free(par_u);
+}
+#endif /* ifdef U3_EPOC_TEST */
+
+int
+main(int argc, char* argv[])
+{
+#ifdef U3_EPOC_TEST
+  mkdir(dir_c, 0777);
+  test_u3_epoc_new();
+  test_sync_mixed();
+  // There's no straightforward way to recursively remove a directory tree from
+  // C, so `rm -rf <dir_c>` will need to be run from the command line to remove
+  // the artifacts from a test run.
+#endif /* ifdef U3_EPOC_TEST */
+
+  fprintf(stderr, "epoc_tests: ok\r\n");
```

Comment:
If these didn't run, because `U3_EPOC_TEST` wasn't defined, then it would be better to print something like `"epoc tests: skipped\r\n"`.

Reply:
> If I `#define U3_EPOC_TEST` in epoc.h, I get this when I run `make test`:
> 
> ```
> build/epoc_tests
> lmdb: failed to open environment at /tmp/event_log/0i138: No such file or directory
> epoc_tests: ok
> ```
> 
> Is that expected?

Yep, it is. The test is ensuring that attempting to open a non-existent epoch fails (see line `76` of `epoc_tests.c`).